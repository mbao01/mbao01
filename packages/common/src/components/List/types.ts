import React from "react";
import { type VariantProps } from "../../libs";
import { getListClasses, getListColumnClasses, getListItemClasses } from "./constants";

export type ListProps = React.HTMLAttributes<HTMLUListElement> &
  VariantProps<typeof getListClasses>;

export type ListItemProps = React.HTMLAttributes<HTMLLIElement> &
  VariantProps<typeof getListItemClasses>;

export type ListColumnProps = React.HTMLAttributes<HTMLLIElement> &
  VariantProps<typeof getListColumnClasses>;
