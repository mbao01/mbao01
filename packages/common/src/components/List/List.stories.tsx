import type { Meta, StoryObj } from "@storybook/react-vite";
import { ReactNode } from "react";
import { List } from "./List";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: List,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    visual: {
      delay: 500, // wait 500ms for test before capturing snapshot
    },
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof List>;

export default meta;
type Story = StoryObj<typeof meta>;

const ListContent: React.FC<{ children?: ReactNode }> = ({ children }) => (
  <List>
    <List.Item>{children}</List.Item>
    <List.Item>{children}</List.Item>
    <List.Item>{children}</List.Item>
  </List>
);

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
  render: () => (
    <ListContent>
      <span>A</span> <span>Nike</span> <span>Shoe!</span>
    </ListContent>
  ),
};

export const LongList: Story = {
  args: {},
  render: () => (
    <ListContent>
      <span>A</span>
      <span>Nike</span>
      <span>Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe!</span>
      <span>Cool!</span>
    </ListContent>
  ),
};

export const ListContentGrow: Story = {
  args: {},
  render: () => (
    <ListContent>
      <span>A</span>
      <span>Nike</span>
      <List.Column flex="grow">
        Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe!
      </List.Column>
      <span>Cool!</span>
    </ListContent>
  ),
};

export const ListContentWrap: Story = {
  args: {},
  render: () => (
    <ListContent>
      <span>A</span>
      <span>Nike</span>
      <List.Column flex="wrap">
        Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe Shoe!
      </List.Column>
      <span>Cool!</span>
    </ListContent>
  ),
};
