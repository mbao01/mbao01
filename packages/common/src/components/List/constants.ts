import { cva } from "../../libs";

export const getListClasses = cva("list");

export const getListItemClasses = cva("list-row");

export const getListColumnClasses = cva("", {
  variants: {
    flex: {
      grow: "list-col-grow",
      wrap: "list-col-grow",
    },
  },
});
