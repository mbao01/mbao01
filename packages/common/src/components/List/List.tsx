import { Slot } from "@radix-ui/react-slot";
import type { ListColumnProps, ListItemProps, ListProps } from "./types";
import { cn } from "../../utilities";
import { getListClasses, getListColumnClasses, getListItemClasses } from "./constants";

const List = ({ className, children, ...props }: ListProps) => (
  <ul className={cn(getListClasses(), className)} {...props}>
    {children}
  </ul>
);

const ListItem = ({ className, children, ...props }: ListItemProps) => (
  <li className={cn(getListItemClasses(), className)} {...props}>
    {children}
  </li>
);

const ListColumn = ({ flex, className, children, ...props }: ListColumnProps) => {
  const SlotChild = ["string", "number", "boolean", "undefined"].includes(typeof children) ? (
    <span>{children}</span>
  ) : (
    children
  );

  return (
    <Slot className={cn(getListColumnClasses({ flex }), className)} {...props}>
      {SlotChild}
    </Slot>
  );
};

List.Item = ListItem;
List.Column = ListColumn;

export { List };
