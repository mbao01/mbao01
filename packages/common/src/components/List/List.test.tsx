import { render, screen } from "@testing-library/react";
import { List } from "./List";

describe("List", () => {
  it("renders a basic list", () => {
    const { asFragment } = render(
      <List>
        <List.Item>
          <span>If a dog chews</span> shoes whose shoes does <span>he choose?</span>
        </List.Item>
      </List>
    );

    expect(screen.getByRole("list")).toBeVisible();
    expect(screen.getByRole("list")).toHaveTextContent(
      "If a dog chews shoes whose shoes does he choose?"
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
