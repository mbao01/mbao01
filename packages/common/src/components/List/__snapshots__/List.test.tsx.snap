// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > renders a basic list 1`] = `
<DocumentFragment>
  <ul
    class="list"
  >
    <li
      class="list-row"
    >
      <span>
        If a dog chews
      </span>
       shoes whose shoes does 
      <span>
        he choose?
      </span>
    </li>
  </ul>
</DocumentFragment>
`;
